Add type descriptions to completions and hovers, support method hovers.

anthony-c-martin opened the pull request:
Closes #668
Closes #2074

![Screen-Recording-2021-04-02-at-1](https://user-images.githubusercontent.com/38542602/113753396-0d987980-96dc-11eb-8c33-b00faf07441a.gif)

* Add type property descriptions to hovers and completions
* Changes to type system & binding so that object properties & methods are dynamically resolved and not bound upfront.

NOTE: This PR does not include an implementation for find references on object property symbols.

anthony-c-martin commented on src/Bicep.Core/Semantics/NameBindingVisitor.cs:
```diff
@@ -224,41 +224,6 @@ public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
             this.bindings.Add(syntax, symbol);
         }
 
-        public override void VisitInstanceFunctionCallSyntax(InstanceFunctionCallSyntax syntax)
```
I'll need to check whether this change breaks find-all-references between `az.someFunc()` and `someFunc()`

anthony-c-martin replied:
I've made the fixes to support this now.

anthony-c-martin commented on src/Bicep.Core/TypeSystem/TypeProperty.cs:
```diff
@@ -5,15 +5,18 @@ namespace Bicep.Core.TypeSystem
     public class TypeProperty
     {
-        public TypeProperty(string name, ITypeReference typeReference, TypePropertyFlags flags = TypePropertyFlags.None)
+        public TypeProperty(string name, ITypeReference typeReference, TypePropertyFlags flags = TypePropertyFlags.None, string? description = null)
```
I'd like to avoid using the default parameter value here, but don't want to add noise to this PR. I may submit this as a follow-up PR.

miqm commented on CONTRIBUTING.md:
```diff
@@ -35,7 +35,7 @@ Many of the bicep integration tests rely on baseline test assertion files that a
 * If you see a test failure with a message containing Windows and *nix copy commands, you have encountered such a test. You have the following options to fix the test:
   1. Manually execute the provided command in a shell. This makes sense for a single test, but is extremely tedious otherwise.
-  1. Run the `SetBaseline.ps1` script at the repo root to execute the tests in `SetBaseLine` mode, which causes the baselines to be automatically updated in bulk for failing tests. You should see baseline file modifications in Git pending changes. (Make sure your Git pending changes are empty before doing so - your changes could get overwritten!).
+  1. Run the `SetBaseline.ps1` script in [scripts](./scripts) to execute the tests in `SetBaseLine` mode, which causes the baselines to be automatically updated in bulk for failing tests. You should see baseline file modifications in Git pending changes. (Make sure your Git pending changes are empty before doing so - your changes could get overwritten!).
```
we could add ``SetBaseline.ps1/`SetBaseline.sh`

anthony-c-martin replied:
Good point!

miqm commented on src/Bicep.Core.Samples/Files/InvalidResources_CRLF/Completions/autoScalerPropertiesRequireEscaping.json:
```diff
@@ -32,7 +32,7 @@
     "documentation": {
       "kind": "markdown",
-      "value": "Type: `'least-waste' | 'most-pods' | 'random'`  \n"
+      "value": "Type: `'least-waste' | 'most-pods' | 'random'`  \n  \n"
```
I assume those are without any description, but are the additional empty lines necessary?

anthony-c-martin replied:
Good catch. Looks like this is coming through from type generation - we legitimately have some properties with description `""`: https://github.com/Azure/bicep-types-az/blob/9db80f44909805c0498e3fb0137fb47f307959f3/generated/containerservice/microsoft.containerservice/2020-09-01/types.md#schemas30_autoscalerprofile

I'll create an issue in the type generation repo to nullify empty/whitespace strings as I don't think it should be Bicep's responsibility to normalize these during type loading.

miqm commented on src/Bicep.Core.Samples/Files/InvalidExpressions_LF/main.diagnostics.bicep:
```diff
@@ -335,7 +335,7 @@ var invalidPropertyCallOnInstanceFunctionAccess = a.b.c.bar().baz
 //@[36:37) [BCP057 (Error)] The name "a" does not exist in the current context. |a|
 var invalidInstanceFunctionCall = az.az()
-//@[37:39) [BCP107 (Error)] The function "az" does not exist in namespace "az". |az|
+//@[37:39) [BCP109 (Error)] The type "az" does not contain function "az". |az|
```
this change in diagnostic is quite odd...

anthony-c-martin replied:
Namespaces are special object types, so this is a result of the change to avoid binding `az.<func>()` at bind time, and instead using dynamic method resolution (which is required for method lookups on all other object types, as they can't be resolved at bind time).

I'll try and special-case this to restore the old error message, as it's more informative.

miqm commented on src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs:
```diff
@@ -89,7 +92,7 @@ public DeclaredTypeManager(IResourceTypeProvider resourceTypeProvider, TypeManag
 
                 case VariableDeclarationSyntax variable:
-                    return new DeclaredTypeAssignment(this.typeManager.GetTypeInfo(variable), variable);
+                    return new DeclaredTypeAssignment(LanguageConstants.Any, variable);
```
Why we're changing this to Any type?

majastrz replied:
Yeah we'll have to revisit this when we add better completions for property access on array or object literals, but that can wait.

shenglol commented on src/Bicep.Core.Samples/Files/InvalidVariables_LF/Completions/objectVarTopLevel.json:
```diff
@@ -1,128 +1 @@
-  }
-]
\ No newline at end of file
+[]
```
This seems odd. Any idea why the completion items are removed?

anthony-c-martin replied:
Good catch. I think because of https://github.com/Azure/bicep/pull/1915#discussion_r610022836. I may need to revise that approach...

majastrz replied:
Yeah...

anthony-c-martin replied:
https://github.com/Azure/bicep/pull/1915/commits/24402525cd45478abfa8f49adec792fbbcb240d4 - here's is a more targeted fix I came up with to avoid the infinite recursion. I don't love it because it feels very special-case.

I did this because:
* language server functionality depends on the declared type for variable declarations being resolved (rather than just `any`)
* we otherwise have infinite recursion (variable declared type -> variable value assigned type -> for loop assigned type -> object assigned type -> object declared type -> for loop declared type -> variable declared type)

shenglol replied:
Personally I prefer
> removing the language server's dependency on declared type

to be the ultimate fix because that avoid special cases, but within the scope of this PR the fix looks good to me.

shenglol replied:
Also I feel like this can be a question for Anders.

shenglol commented on src/Bicep.Core.UnitTests/Utils/TestSyntaxHelper.cs:
```diff
@@ -11,19 +11,14 @@ public static class TestSyntaxHelper
+                return false;
             }
 
-            if (instanceFunctionCallSyntax.BaseExpression is VariableAccessSyntax baseVariableSyntax && 
-                semanticModel.Root.ImportedNamespaces.ContainsKey(baseVariableSyntax.Name.IdentifierName))
-            {
-                // we only expect to have bound InstanceFunctionCallsSyntax if they accessed on a namespace - e.g. sys.concat(..)
-                return true;
-            }
-
-            return false;
+            return true;
         }
```
[nit] The method body can be simplified to
```
=> symbolReference is not InstanceFunctionCallSyntax and not PropertyAccessSyntax and not ObjectPropertySyntax;
```

shenglol commented on src/Bicep.LangServer.IntegrationTests/DefinitionTests.cs:
```diff
@@ -81,7 +83,7 @@ public async Task GoToDefinitionRequestOnUnsupportedOrInvalidSyntaxNodeShouldRet
             var lineStarts = compilation.SyntaxTreeGrouping.EntryPoint.LineStarts;
 
-            var undeclaredSymbolBindings = symbolTable.Where(pair => !(pair.Value is DeclaredSymbol));
+            var undeclaredSymbolBindings = symbolTable.Where(pair => pair.Value is not DeclaredSymbol && pair.Value is not PropertySymbol);
```
[nit]
```suggestion
            var undeclaredSymbolBindings = symbolTable.Where(pair => pair.Value is not and not PropertySymbol);
```

majastrz commented on src/Bicep.Core/TypeSystem/FunctionFlags.cs:
```diff
@@ -51,7 +51,12 @@ public enum FunctionFlags
 
         /// <summary>
-        /// The function can be used as an output decorator.
+        /// The function can be used as any decorator.
```
>any decorator [](start = 40, length = 13)

Do you mean "can be used as a decorator anywhere"?

majastrz commented on src/Bicep.Core.Samples/Files/InvalidVariables_LF/main.diagnostics.bicep:
```diff
@@ -132,9 +132,9 @@ var objWithInterp = {
+//@[15:18) [BCP109 (Error)] The type "az" does not contain function "add". |add|
 var myConcat = sys.concat('a', az.concat('b', 'c'))
-//@[34:40) [BCP107 (Error)] The function "concat" does not exist in namespace "az". |concat|
+//@[34:40) [BCP109 (Error)] The type "az" does not contain function "concat". |concat|
```
>The [](start = 28, length = 3)

Can we special case this and restore the original message? As a user unfamiliar with the internals, I would have no idea that "type az" means namespace.

anthony-c-martin replied:
Done

majastrz commented on src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs:
```diff
@@ -514,13 +514,26 @@ public override void VisitObjectSyntax(ObjectSyntax syntax)
+                            return new TypeProperty(property.Name, resolvedType, property.Flags, property.Description);
+                        }
+
+                        return new TypeProperty(group.Key, resolvedType);
```
>TypeProperty [](start = 35, length = 12)

Can you add comments explaining each branch? I'm guessing L532 covers the cases where we have a property that doesn't exist in the types?

majastrz commented on src/Bicep.Core/Semantics/SemanticModel.cs:
```diff
@@ -114,14 +115,85 @@ public bool HasErrors()
+        {
+            static PropertySymbol? GetPropertySymbol(TypeSymbol? baseType, string property)
+            {
+                var bodyObjectType = baseType switch
```
>switch [](start = 46, length = 6)

Can you reuse `TypeAssignmentVisitor.UnwrapType()` here?

anthony-c-martin replied:
Yup! I looked for that function but didn't find it.

majastrz commented on src/Bicep.Core/Syntax/SyntaxVisitor.cs:
```diff
@@ -15,6 +16,8 @@ public void Visit(SyntaxBase? node)
                 return;
             }
 
+            RuntimeHelpers.EnsureSufficientExecutionStack();
```
>EnsureSufficientExecutionStack [](start = 27, length = 30)

Nice! (I think there is an issue tracking this.)

majastrz commented on src/Bicep.Core.IntegrationTests/Emit/TemplateEmitterTests.cs:
```diff
@@ -37,8 +37,8 @@ public void ValidBicep_TemplateEmiterShouldProduceExpectedTemplate(DataSet dataS
             // emitting the template should be successful
             var result = this.EmitTemplate(SyntaxTreeGroupingBuilder.Build(new FileResolver(), new Workspace(), PathHelper.FilePathToFileUrl(bicepFilePath)), compiledFilePath, BicepTestConstants.DevAssemblyFileVersion);
-            result.Status.Should().Be(EmitStatus.Succeeded);
             result.Diagnostics.Should().BeEmptyOrContainDeprecatedDiagnosticOnly();
```
>BeEmptyOrContainDeprecatedDiagnosticOnly [](start = 40, length = 40)

Yeah good pt. The reordering produces better error messages on failures.

majastrz reviewed (changes requested):
Added some comments.

majastrz commented:
I tried getting a hover for a discriminator property and didn't get anything:
![image](https://user-images.githubusercontent.com/22460039/114109285-d36cda80-9889-11eb-83e0-bb46734b5962.png)

But it worked for a regular property:
![image](https://user-images.githubusercontent.com/22460039/114109312-e2ec2380-9889-11eb-9ee8-69c0b869cb71.png)

Same problem in a non top-level discriminator property as well:
![image](https://user-images.githubusercontent.com/22460039/114109424-13cc5880-988a-11eb-9a97-a669551d9234.png)

It may be difficult to produce a description for that property depending on how the swagger is structured for that type, but we should show something similar to what the completion text looks like for the same property.

majastrz commented:
Btw, I'm assuming that find refs for properties is out of scope of this PR?

anthony-c-martin commented:
> Btw, I'm assuming that find refs for properties is out of scope of this PR?

Yes. I've added that as a note to the description. This implementation simply generates a new PropertySymbol when requested with no caching. Find refs would need a cache layer, plus means answering some tough questions about what we consider equivalent when comparing properties.

anthony-c-martin commented:
> I tried getting a hover for a discriminator property and didn't get anything:

Good find! I've fixed the hover and added a test for it.

As you mention, showing a description is a little tricky. I think we have all the information we need in swagger, but we currently store descriptions in the individual property fields - meaning we only have descriptions for the sub class discriminator properties and not the base class discriminator property. I'll add an issue to the types repo for this.

shenglol reviewed (approved):
This is awesome!

majastrz reviewed (approved):
This is great!

miqm commented:
Seems like #1571 is unblocked, I’ll bring my branch up2date sometime next week.

anthony-c-martin commented:
> Seems like #1571 is unblocked, I’ll bring my branch up2date sometime next week.

I was actually partway through merging main into your branch; I'll send you a PR if I get there, so you can choose whether or not you want to accept it.

miqm commented:
> > Seems like #1571 is unblocked, I’ll bring my branch up2date sometime next week.
> 
> I was actually partway through merging main into your branch; I'll send you a PR if I get there, so you can choose whether or not you want to accept it.

awesome! thanks :)

anthony-c-martin commented:
> awesome! thanks :)

@miqm I pushed accidentally rather than submitting a PR. Tests are passing though, which is a good sign!
